include/foxy/detail/tunnel.hpp

```diff
@@ -97,51 +108,207 @@ struct tunnel_op : boost::asio::coroutine
              bool const                is_continuation = true) -> void;
+
+  auto
+  operator()(on_connect_t, boost::system::error_code ec, boost::asio::ip::tcp::endpoint) -> void;
+
+  auto
+  operator()(on_relay_t, boost::system::error_code ec, bool close_tunnel) -> void;
 };
 
-template <class Stream, class TunnelHandler>
+template <class TunnelHandler>
+auto
+tunnel_op<TunnelHandler>::
+operator()(on_connect_t, boost::system::error_code ec, boost::asio::ip::tcp::endpoint) -> void
+{
+  (*this)(ec, 0);
+}
+
+template <class TunnelHandler>
+auto
+tunnel_op<TunnelHandler>::operator()(on_relay_t, boost::system::error_code ec, bool) -> void
+{
+  (*this)(ec, 0);
+}
+
+template <class TunnelHandler>
 auto
-tunnel_op<Stream, TunnelHandler>::
-operator()(boost::system::error_code ec,
-           std::size_t const         bytes_transferred,
-           bool const                is_continuation) -> void
+tunnel_op<TunnelHandler>::operator()(boost::system::error_code ec,
+                                     std::size_t               bytes_transferred,
+                                     bool const                is_continuation) -> void
 {
   using namespace std::placeholders;
   using boost::beast::bind_handler;
 
+  namespace net  = boost::asio;
   namespace http = boost::beast::http;
 
   auto& s = *p_;
   BOOST_ASIO_CORO_REENTER(*this)
   {
+    while (true) {
+      s.parser.emplace(std::piecewise_construct, std::make_tuple(),
+                       std::make_tuple(get_allocator()));
+
+      s.response.emplace(std::piecewise_construct, std::make_tuple(get_allocator()),
+                         std::make_tuple(get_allocator()));
+
+      std::cout << "reading in client header...\n";
+
+      BOOST_ASIO_CORO_YIELD
+      s.server.async_read_header(*s.parser, std::move(*this));
+
+      if (ec) { goto upcall; }
+
+      std::cout << "parsing URI now:" << s.parser->get().target() << "\n";
+
+      s.uri_parts = foxy::parse_uri(s.parser->get().target());
+
+      s.is_authority = s.uri_parts.is_authority();
+      s.is_connect   = s.parser->get().method() == http::verb::connect;
+      s.is_absolute  = s.uri_parts.is_absolute();
+      s.is_http      = s.uri_parts.is_http();
+
+      if (s.is_connect && s.is_authority && !s.parser->keep_alive()) {
+        std::cout << "sending error response...\n";
+
+        s.close_tunnel = true;
+
+        s.response->result(http::status::bad_request);
+        s.response->body() = "CONNECT semantics require a persistent connection\n\n";
+        s.response->prepare_payload();
+
+        BOOST_ASIO_CORO_YIELD
+        s.server.async_write(*s.response, std::move(*this));
+
+        if (ec) { goto upcall; }
+        break;
+      }
+
+      if ((s.is_connect && s.is_authority) || (s.is_absolute && s.is_http)) {
+        std::cout << "connecting to: \n" << s.uri_parts.host() << "\n\n";
+
+        BOOST_ASIO_CORO_YIELD
+        {
+          auto const scheme = s.uri_parts.scheme().size() == 0
```

Comment:
Business logic like this should be in its own function that deals with messages only